presto-main/src/main/java/com/facebook/presto/cost/CostComparator.java

```diff
@@ -58,6 +63,16 @@ public int compare(Session session, PlanNodeCostEstimate left, PlanNodeCostEstim
         requireNonNull(left, "left can not be null");
         requireNonNull(right, "right can not be null");
         checkArgument(!left.isUnknown() && !right.isUnknown(), "cannot compare unknown costs");
+
+        long memoryLimit = getMemoryLimit(session);
+
+        if (memoryLimit < left.getMemoryCost() && memoryLimit >= right.getMemoryCost())  {
```

Comment:
Isn't the estimate for memory for a PlanNode the amount it can take on a particular node or within a fragment?